crates/assets/src/asset.rs

```diff
@@ -0,0 +1,949 @@
+	/// The minimum length of a valid asset ID identifier.
+	pub const MINIMUM_IDENTIFIER_LENGTH: usize = 1;
+	/// The maximum length of a valid asset ID reference.
+	pub const MAXIMUM_IDENTIFIER_LENGTH: usize = 78;
+	const MAXIMUM_IDENTIFIER_LENGTH_U32: u32 = MAXIMUM_IDENTIFIER_LENGTH as u32;
+
+	/// Separator between asset namespace and asset reference.
+	const NAMESPACE_REFERENCE_SEPARATOR: u8 = b':';
+	/// Separator between asset reference and asset identifier.
+	const REFERENCE_IDENTIFIER_SEPARATOR: u8 = b':';
+
+	/// Namespace for Slip44 assets.
+	pub const SLIP44_NAMESPACE: &[u8] = b"slip44";
+	/// Namespace for Erc20 assets.
+	pub const ERC20_NAMESPACE: &[u8] = b"erc20";
+	/// Namespace for Erc721 assets.
+	pub const ERC721_NAMESPACE: &[u8] = b"erc721";
+	/// Namespace for Erc1155 assets.
+	pub const ERC1155_NAMESPACE: &[u8] = b"erc1155";
+
+	// TODO: Add link to the Asset DID spec once merged.
+
+	/// The Asset ID component as specified in the Asset DID specification.
+	#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub enum AssetId {
+		// A SLIP44 asset reference.
+		Slip44(Slip44Reference),
+		// An ERC20 asset reference.
+		Erc20(EvmSmartContractFungibleReference),
+		// An ERC721 asset reference.
+		Erc721(EvmSmartContractNonFungibleReference),
+		// An ERC1155 asset reference.
+		Erc1155(EvmSmartContractNonFungibleReference),
+		// A generic asset.
+		Generic(GenericAssetId),
+	}
+
+	impl From<Slip44Reference> for AssetId {
+		fn from(reference: Slip44Reference) -> Self {
+			Self::Slip44(reference)
+		}
+	}
+
+	impl From<EvmSmartContractFungibleReference> for AssetId {
+		fn from(reference: EvmSmartContractFungibleReference) -> Self {
+			Self::Erc20(reference)
+		}
+	}
+
+	impl AssetId {
+		/// Try to parse an `AssetId` instance from the provided UTF8-encoded
+		/// input.
+		pub fn from_utf8_encoded<I>(input: I) -> Result<Self, Error>
+		where
+			I: AsRef<[u8]> + Into<Vec<u8>>,
+		{
+			let input = input.as_ref();
+			let input_length = input.len();
+			if !(MINIMUM_ASSET_ID_LENGTH..=MAXIMUM_ASSET_ID_LENGTH).contains(&input_length) {
+				log::trace!(
+					"Length of provided input {} is not included in the inclusive range [{},{}]",
+					input_length,
+					MINIMUM_ASSET_ID_LENGTH,
+					MAXIMUM_ASSET_ID_LENGTH
+				);
+				return Err(Error::InvalidFormat);
+			}
+
+			let AssetComponents {
+				namespace,
+				reference,
+				identifier,
+			} = split_components(input);
+
+			match (namespace, reference, identifier) {
+				// "slip44:" assets -> https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-20.md
+				(Some(SLIP44_NAMESPACE), Some(slip44_reference), identifier) => {
+					if identifier.is_some() {
+						log::trace!("Slip44 namespace does not accept an asset identifier.");
+						Err(Error::InvalidFormat)
+					} else {
+						Slip44Reference::from_utf8_encoded(slip44_reference).map(Self::Slip44)
+					}
+				}
```

Comment:
So having two matches for each case? If that is your suggestion, I think it makes it more confusing because we already have 4 cases, some of which might have an identifier component, some of which may not. This way I thought it would help to separate better the different cases.

Reply:
What's your take @weichweich ?

Reply:
Fixed that in https://github.com/KILTprotocol/kilt-node/pull/378/commits/6e54dd9f793a1f40a84627b26144e351124eaf0a.